include/obmc_hypervisor.hpp

```diff
@@ -0,0 +1,161 @@
+#pragma once
+#include <sys/socket.h>
+
+#include <app.hpp>
+#include <async_resp.hpp>
+#include <boost/asio/local/stream_protocol.hpp>
+#include <boost/container/flat_map.hpp>
+#include <boost/container/flat_set.hpp>
+#include <websocket.hpp>
+
+#include <array>
+#include <memory>
+#include <string>
+
+namespace crow
+{
+namespace obmc_hypervisor
+{
+
+// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)
+static std::unique_ptr<boost::asio::local::stream_protocol::socket> hostSocket;
+
+static std::array<char, 4096> outputBuffer;
+static std::string inputBuffer;
+
+static boost::container::flat_set<crow::websocket::Connection*> sessions;
+
+static bool doingWrite = false;
+// NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
+
+inline void doWrite()
+{
+    if (doingWrite)
+    {
+        BMCWEB_LOG_DEBUG("Already writing.  Bailing out");
+        return;
+    }
+
+    if (inputBuffer.empty())
+    {
+        BMCWEB_LOG_DEBUG("Outbuffer empty.  Bailing out");
+        return;
+    }
+
+    if (!hostSocket)
+    {
+        BMCWEB_LOG_ERROR("doWrite(): Socket closed.");
+        return;
+    }
+
+    doingWrite = true;
+    hostSocket->async_write_some(
+        boost::asio::buffer(inputBuffer.data(), inputBuffer.size()),
+        [](const boost::beast::error_code& ec, std::size_t bytesWritten) {
+        doingWrite = false;
+        inputBuffer.erase(0, bytesWritten);
+
+        if (ec == boost::asio::error::eof)
```

Comment:
I think this should be checking against ec.value().

Reply:
I think - the current comparison (e.g. `if(ec.value() == EBADR)`) is actually semantically incorrect.

- `error_code` is a structure with `value` & `category`, and the value may potentially the same if the categories are different.
- In current bmcweb code (and most cases), it assumes they are in `system_category` and thus the comparison with `ec.value()` may be valid.

The code `if (ec == boost::asio::error::eof)` is translated to `if(ec == std::make_error_condition(boost::asio::error::eof)` and it is doing the comparison correctly.

https://www.boost.org/doc/libs/1_84_0/boost/system/detail/error_code.hpp
```
    template<class E, class N = typename detail::enable_if<std::is_error_condition_enum<E>::value>::type>
    inline friend bool operator==( error_code const & lhs, E rhs ) BOOST_NOEXCEPT
    {
        return lhs == make_error_condition( rhs );
    }
```

In case the rhs value is for system error (e.g. EBADR), the semantically-correct way would be

`if(ec.value() == EBADR)`
-->
`if (ec == boost::system::error_code{EBADR, boost::system::system_category()})`


Ideally, the use of the direct comparison of 'ec.value()` may need to be avoided unless the error category is known.

Reply:
Thank you for the explanation. I see what you mean and agree with you.